### PR TITLE
Refactor insert high school classes and insert statistics from 2020

### PR DESCRIPTION
**Pull request by konradkrasno**
I replaced 0012_insert_profile_2020.py with 0012_insert_profile_2020_and_scores.py. I moved functions responsible for adding data to search/load_data.py.
There are two functions for load data from publiczne_oddzialy_2020.csv and scores_2020.csv (file with punctations data). There are some issues for discussion:
- if a school with provided REGON does not exist in db and we want to create a new one what to do with missing data?
- I change the name of HighSchoolClass to concatinated string "Symbol oddziału" and "Nazwa krótka oddziału". I use it to bind class models with punctation data. Please let me know if this does not create problems at the front.
- I think there is no need to create new models for technikum and vocational classes. I think one model for classes will be enough. IMO it is sufficient to add a type of school to the existing model if this information is needed directly from the class model (this information is also contained in the related school model.)
- there is also an issue with regex for single language abbr, e.g.: (ang) -> raise NoneType Error

@annkamsk @micorix  please for review

### CODE REVIEW

**Comment by micorix** [comment]
Ad 1.
a) I'd def add schools with missing data. We don't want to loose sth. (All data from files we get from the city should be always stored in db - imo it's a good rule, lmk if u disagree)
b) @annkamsk i know u already has written (and it's merged) migration for inserting classes data but wouldn't it be better to write migrations for inserting school info first in the future? That way, we can handle all mess regarding missing REGONs ealier - REGON is part of school info, classes data may have it's own issues, so I wouldn't try to squash it into one huge PR. idk just a thought
c) Maybe some REGON-fuzzy search algorithm would be also helpful. We would need to validate it manually (?) but still.
d) Maybe it's time come up with some API-wide value for N/A data (not 0 - zero is invalid regon, but it could be a valid value for some other field)? (so that frontend can pick that up and display some popup "missing data, would u be so kind to.. etc.)

File: search/load_data.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            #     institution_nr=item.get(""),  # no data!
+            #     institution_RSPO=item.get(""),  # no data!
+            #     RSPO=item.get(""),  # no data!
+            #     institution_regon=item.get(""),  # no data!
+            #     regon=item.get(""),  # no data!
+            #     data=item.get(""),  # no data!
+            # )
+            school = School.objects.create(
+                school_name=item.get("Nazwa szkoły"),
+                school_type=item.get("Typ szkoły (LO/LP/TECH/BR1St/LU/TU/PS)"),
```
Comment (annkamsk) [inline search/load_data.py:71]: This should already be specified in code a long time ago, but the school_type can be:
przedszkole, liceum ogólnokształcące, technikum, szkoła podstawowa, szkoła policealna. 
School type generalized can be:
przedszkole, szkoła ponadpodstawowa, szkoła policealna, szkoła podstawowa

**Review by annkamsk — CHANGES_REQUESTED** [review]
Some general comments:
1. Don’t change already existing migration files. Doing this doesn’t revert the database to previous state – it only removes the code of the migration, making it hard to track the changes that happened to db. We could potentially manually revert the migration after merge, but only if there is an important reason for that. 
2. For this reason it doesn’t really make sense to leave TODOs in migration files (unless you want to fix them before merging the PR, but in this case it’s better to create a draft pull request: https://github.blog/2019-02-14-introducing-draft-pull-requests/)

**Comment by annkamsk** [comment]
> b) @annkamsk i know u already has written (and it's merged) migration for inserting classes data but wouldn't it be better to write migrations for inserting school info first in the future? That way, we can handle all mess regarding missing REGONs ealier - REGON is part of school info, classes data may have it's own issues, so I wouldn't try to squash it into one huge PR. idk just a thought
> c) Maybe some REGON-fuzzy search algorithm would be also helpful. We would need to validate it manually (?) but still.
> d) Maybe it's time to come up with some API-wide value for N/A data (not 0 - zero is invalid regon, but it could be a valid value for some other field)? (so that frontend can pick that up and display some popup "missing data, would u be so kind to.. etc.)

b) In general, it’s a better idea to put the schools first. In this case I decided to go with classes first, because almost all of the schools were already in db, and having data on classes was much more important than adding few missing schools. 
And squashing these changes into one PR doesn’t make sense because we already have most of the data. Instead, we should make a migration that creates the schools that were missing and add classes to them
c) No need. Very few schools have missing regon. Manual check is enough.
d) Yeah, it’s called a null :P We can’t use it on many of the fields now, because they were created as non-nullable. We need to change it before inserting the data

**Comment by micorix** [comment]
Yeah, null lol

**Comment by annkamsk** [comment]
I made some small changes to School model in #34 that should make the work easier

File: search/load_data.py (hunk 1 above)
Comment (annkamsk) [reply search/load_data.py:71]: After #34 the school_type has choices set and school_type_generlized is removed

File: search/load_data.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                school_type_generalised="fake",  # TODO create mapping based on school_type
```
[the remark this answers is not in the document]
Comment (konradkrasno) [reply search/load_data.py:72]: I did not want to change models without prior arrangement. This is only draft.

File: requirements.txt [hunk 3]
```diff
@@ -26,3 +26,6 @@ uritemplate==3.0.1
 urllib3==1.25.8
 uwsgi==2.0.18
 whitenoise==5.0.1
+pandas==1.2.4
+pytest==6.2.3
```
[the remark this answers is not in the document]
Comment (konradkrasno) [reply requirements.txt:30]: In my opinion it's better than unittest

**Comment by konradkrasno** [comment]
So, I restore 0012_insert_profile_2020.py and create new migration for updating data about schools and inserting scores. Is that ok?

File: requirements.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (annkamsk) [reply requirements.txt:30]: But why? :P

**Comment by annkamsk** [comment]
> So, I restore 0012_insert_profile_2020.py and create new migration for updating data about schools and inserting scores. Is that ok?

Sounds good to me :) Remember to rebase to current master first, because I added #34 and #35, which are gonna be useful

**Comment by konradkrasno** [comment]
@annkamsk please review last commited changes

File: search/migrations/0015_update_schools_insert_scores_2020.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        try:
+            school = School.objects.get(public_institution_data__regon=regon)
+        except School.DoesNotExist:
+            continue
+        else:
+            school.address.city = item.get("Miejscowość szkoły")
+            school.address.postcode = item.get("Kod pocztowy szkoły")
+            school.address.district = item.get("Dzielnica")
+            school.address.street = item.get("Ulica szkoły")
+            school.address.building_nr = item.get("Nr budynku szkoły")
```
Comment (annkamsk) [inline search/migrations/0015_update_schools_insert_scores_2020.py:21]: you need to call school.address.save() at the end

File: search/migrations/0015_update_schools_insert_scores_2020.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            school.school_name = item.get("Nazwa szkoły").strip()
+            school.school_type = item.get("Typ szkoły (LO/LP/TECH/BR1St/LU/TU/PS)")
+            school.is_public = item.get("Uprawnienia publiczne") == "szkoła publiczna"
```
Comment (annkamsk) [inline search/migrations/0015_update_schools_insert_scores_2020.py:24]: you need to call school.save()

File: search/migrations/0015_update_schools_insert_scores_2020.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        except HighSchoolClass.DoesNotExist:
+            continue
+        else:
+            Statistics.objects.create(
+                high_school_class=hsc,
+                points_min=float(item.get("points_min").replace(",", ".")),
+                points_max=float(item.get("points_max").replace(",", ".")),
+                points_avg=float(item.get("points_avg").replace(",", ".")),
+                round=1,
+            )
```
Comment (annkamsk) [inline search/migrations/0015_update_schools_insert_scores_2020.py:48]: You can also update the HighSchoolClass.name here to what is set in „school_class”

File: search/migrations/0015_update_schools_insert_scores_2020.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def load_data(apps, schema_editor):
+    update_school_data(apps)
+    insert_scores(apps)
+
+
+def reverse(apps, schema_editor):
+    HighSchoolClass = apps.get_model("search", "HighSchoolClass")
+    HighSchoolClass.objects.filter(year=(2020, 2022)).delete()
```
Comment (annkamsk) [inline search/migrations/0015_update_schools_insert_scores_2020.py:58]: You’d rather want to remove Statistics for these classes, not the classes themselves (they were loaded in other migration).

**Review by annkamsk — CHANGES_REQUESTED** [review]

**Review by annkamsk — APPROVED** [review]

**Review by mlazowik — APPROVED** [review]

**Pull request merged**